File: FieldKit.Modules.Products/PriceResolver.cs
using System.Text.Json.Serialization;

namespace FieldKit.Modules.Products;

/// <summary>Which kind of assignment a resolved price came from.</summary>
/// <remarks>
/// <para>
/// Returned rather than inferred, so a caller can explain the answer. "Why is this shop paying
/// this?" is the question a field rep asks a supervisor, and the specificity is most of the answer.
/// </para>
/// <para>
/// <b>Ordered least-specific first, and <see cref="PriceResolver"/> compares these values with
/// <c>&gt;</c>.</b> That is a real coupling and it is why the numbers are written out: a scope
/// inserted in the middle would silently re-rank every price in the system, with no test failing
/// except the vectors. A new scope goes at the end if it is more specific than <see cref="Outlet"/>,
/// and otherwise the members get renumbered <i>and</i> the wire format checked — these cross the API
/// by name (<c>ResolvedPriceResponse.Scope</c>), so renumbering is safe for clients but never for
/// stored ordinals.
/// </para>
/// </remarks>
[JsonConverter(typeof(JsonStringEnumConverter<PriceScope>))]
public enum PriceScope
{
    /// <summary>Set for every outlet trading in a channel.</summary>
    Channel = 0,

    /// <summary>Set for one shop, deliberately — an exception to its channel's price.</summary>
    Outlet = 1,
}

/// <summary>
/// One price this product could have on this date, and where it came from.
/// </summary>
/// <remarks>
/// Deliberately flat and self-contained: no ids to follow, no entities, nothing to load. That is
/// what lets <see cref="PriceResolver"/> be a pure function over data the caller has already
/// gathered — and what lets the same shape exist in TypeScript for the device mirror (<c>PRD-08</c>)
/// without dragging an ORM behind it.
/// </remarks>
/// <param name="Amount">Net, in <paramref name="Currency"/>. A decimal, never a float (<c>BR-PRD-8</c>).</param>
public sealed record PriceCandidate(
    Guid PriceListId,
    PriceScope Scope,
    string Currency,
    DateOnly EffectiveFrom,
    DateOnly? EffectiveTo,
    decimal Amount);

/// <summary>The price that applies, and which candidate won.</summary>
public sealed record ResolvedPrice(
    Guid PriceListId, PriceScope Scope, string Currency, decimal Amount);

/// <summary>
/// Picks the price that applies to one product at one outlet on one date (<c>PRD-04</c>,
/// <c>BR-PRD-2</c>).
/// </summary>
/// <remarks>
/// <para>
/// <b>Pure and side-effect-free</b> (<c>BR-PRD-7</c>): candidates in, one answer out. No database, no
/// clock, no tenant context, no logging. That is not tidiness — it is the requirement that makes the
/// same rules runnable on a device that is offline, and testable against vectors a TypeScript
/// implementation must also satisfy (<c>PRD-08</c>).
/// </para>
/// <para>
/// The <i>date</i> is a parameter rather than read from a clock for the same reason. Resolution has
/// to be reproducible: the price an order was placed at must still resolve to that price when the
/// order syncs three days later, and a function that asks what day it is cannot promise that.
/// </para>
/// </remarks>
public static class PriceResolver
{
    /// <summary>
    /// The applicable price, or null when nothing covers <paramref name="on"/>.
    /// </summary>
    /// <remarks>
    /// <para>
    /// <b>Null is a real answer.</b> A product with no price list covering the date is not an error
    /// here — it is a product this outlet cannot be sold, which is a decision for the caller
    /// (<c>BR-PRD-4</c> territory) rather than an exception thrown from inside a pure function.
    /// </para>
    /// <para>
    /// Order of preference, per <c>BR-PRD-2</c>:
    /// </para>
    /// <list type="number">
    /// <item>candidates whose window covers the date — everything else is not a candidate at all;</item>
    /// <item><see cref="PriceScope.Outlet"/> beats <see cref="PriceScope.Channel"/>, always, even
    /// when the channel list starts later. A price set for one shop is a deliberate exception, and a
    /// newer channel-wide list should not silently overwrite it;</item>
    /// <item>within one scope, the latest <c>EffectiveFrom</c> wins — the most recent decision;</item>
    /// <item>still tied, the higher <c>PriceListId</c> wins, ordered as bytes. See below.</item>
    /// </list>
    /// </remarks>
    public static ResolvedPrice? Resolve(IReadOnlyList<PriceCandidate> candidates, DateOnly on)
    {
        ResolvedPrice? best = null;
        PriceCandidate? winner = null;

        foreach (var candidate in candidates)
        {
            if (!Covers(candidate, on)) continue;
            if (winner is not null && !Beats(candidate, winner)) continue;

            winner = candidate;
            best = new ResolvedPrice(
                candidate.PriceListId, candidate.Scope, candidate.Currency, candidate.Amount);
        }

        return best;
    }

    /// <summary>Half-open: <c>[EffectiveFrom, EffectiveTo)</c>, matching <see cref="PriceList"/>.</summary>
    private static bool Covers(PriceCandidate candidate, DateOnly on) =>
        on >= candidate.EffectiveFrom
        && (candidate.EffectiveTo is not { } end || on < end);

    /// <summary>
    /// Whether <paramref name="challenger"/> should displace <paramref name="holder"/>.
    /// </summary>
    /// <remarks>
    /// <para>
    /// <b>The last comparison is the one worth explaining.</b> Two lists at the same scope with the
    /// same effective date is a data problem — an author has said two contradictory things — and no
    /// tiebreak makes it *right*. What a tiebreak buys is determinism: the same inputs give the same
    /// answer on the server and on every device, so a rep and a supervisor looking at the same shop
    /// see the same number, and an order re-priced during sync does not change.
    /// </para>
    /// <para>
    /// Comparing ids gives that. Ids here are UUIDv7, which are creation-ordered, so the higher id is
    /// the more recently authored list — the same instinct as the effective-date rule above, applied
    /// one level down.
    /// </para>
    /// <para>
    /// <b>Ordered as big-endian bytes, which is what the canonical string spells out.</b> The rule
    /// has to be stateable in a sentence both languages can implement, because the mirror is
    /// TypeScript and has no <c>Guid</c> type at all — "compare the 16 bytes in the order the
    /// canonical form prints them" is implementable anywhere; "whatever this platform's Guid
    /// comparison does" is not a specification.
    /// </para>
    /// <para>
    /// <b>The trap this avoids is <c>Guid.ToByteArray()</c>, not <c>Guid.CompareTo</c>.</b> The
    /// no-argument overload returns the first three groups <i>little-endian</i>, so comparing those
    /// bytes disagrees with the canonical string: <c>00000100-…</c> sorts below <c>00000002-…</c>,
    /// backwards. That is the mistake a C# implementer actually makes, and the vectors carry a pair
    /// that catches it. <c>Guid.CompareTo</c> on this runtime compares those fields as
    /// <i>unsigned</i> and agrees with byte order — an earlier version of this comment claimed
    /// otherwise and was wrong; see <c>vectors/README.md</c>.
    /// </para>
    /// </remarks>
    private static bool Beats(PriceCandidate challenger, PriceCandidate holder)
    {
        if (challenger.Scope != holder.Scope) return challenger.Scope > holder.Scope;

        if (challenger.EffectiveFrom != holder.EffectiveFrom)
        {
            return challenger.EffectiveFrom > holder.EffectiveFrom;
        }

        Span<byte> left = stackalloc byte[16];
        Span<byte> right = stackalloc byte[16];
        challenger.PriceListId.TryWriteBytes(left, bigEndian: true, out _);
        holder.PriceListId.TryWriteBytes(right, bigEndian: true, out _);

        return left.SequenceCompareTo(right) > 0;
    }
}

File: frontend/lib/pricing/price-resolver.ts
/**
 * Which kind of assignment a resolved price came from.
 *
 * Names, not numbers — the same values the API and the vector files carry. The C# enum's ordinals
 * are storage; what crosses a boundary is the word.
 */
export type PriceScope = "Channel" | "Outlet";

/**
 * How specific each scope is. **`Outlet` beats `Channel`, always.**
 *
 * A rank table rather than a relied-upon declaration order, because TypeScript string unions have no
 * ordinal to compare and inventing one from array position would be the same silent coupling the C#
 * enum's comment warns about. A scope added here has to be given a rank deliberately.
 */
const SPECIFICITY: Readonly<Record<PriceScope, number>> = {
  Channel: 0,
  Outlet: 1,
};

/**
 * One price this product could have on this date, and where it came from.
 *
 * **`amount` is the string the caller was given, and stays one.** Resolution is a *selection* rule:
 * it picks a candidate, it does not do arithmetic. Parsing the amount here would mean re-formatting
 * it on the way out, and `"12.50"` becoming `"12.5"` is a different answer to give a rep even though
 * it is the same number. The arithmetic engines (tax, promotions) take it from here as a `Money`.
 */
export type PriceCandidate = {
  priceListId: string;
  scope: PriceScope;
  currency: string;
  /** `YYYY-MM-DD`. A business day, not an instant — see `resolvePrice`. */
  effectiveFrom: string;
  effectiveTo: string | null;
  amount: string;
};

/** The price that applies, and which candidate won. */
export type ResolvedPrice = {
  priceListId: string;
  scope: PriceScope;
  currency: string;
  amount: string;
};

/**
 * Picks the price that applies to one product at one outlet on one date (`PRD-04`, `BR-PRD-2`) —
 * the device's mirror of `PriceResolver.Resolve`.
 *
 * **Pure** (`BR-PRD-7`): candidates in, one answer out. No fetch, no clock, no storage. That is what
 * lets it run on a phone with no signal, and what lets the same vector file check both languages
 * (`PRD-08`).
 *
 * **The date is a parameter, and a string.** Reproducibility is the reason for the first: an order
 * placed at a price must still resolve to that price when it syncs three days later, and a function
 * that asks what day it is cannot promise that. A `YYYY-MM-DD` string rather than a `Date` is the
 * reason for the second — a business day is not an instant (`BR-PRD-6`), and `new Date("2026-03-15")`
 * is midnight *UTC*, which is the previous day in Bucharest for anyone west of it. Comparing the
 * strings is exact, timezone-free, and gives the same answer as comparing the days.
 *
 * **Null is a real answer**, not an error: a product with no list covering the date is one this
 * outlet cannot be sold, which is the caller's decision to make.
 */
export function resolvePrice(
  candidates: readonly PriceCandidate[],
  on: string,
): ResolvedPrice | null {
  let winner: PriceCandidate | null = null;

  for (const candidate of candidates) {
    if (!covers(candidate, on)) continue;
    if (winner !== null && !beats(candidate, winner)) continue;

    winner = candidate;
  }

  return winner === null
    ? null
    : {
        priceListId: winner.priceListId,
        scope: winner.scope,
        currency: winner.currency,
        amount: winner.amount,
      };
}

/** Half-open: `[effectiveFrom, effectiveTo)` — a successor starts the day its predecessor stops. */
function covers(candidate: PriceCandidate, on: string): boolean {
  return on >= candidate.effectiveFrom
    && (candidate.effectiveTo === null || on < candidate.effectiveTo);
}

/**
 * Whether `challenger` should displace `holder` (`BR-PRD-2`).
 *
 * In order: a more specific scope wins; then the later `effectiveFrom` — the most recent decision;
 * then the higher id.
 *
 * **The id tiebreak is not a rule about correctness, it is a rule about agreement.** Two lists at the
 * same scope with the same start date is a data problem, and no tiebreak makes it right. What it buys
 * is determinism: the server and every device pick the same one, so a rep and a supervisor see the
 * same number and an order re-priced during sync does not move.
 *
 * **Compared as lowercase canonical strings**, which `vectors/README.md` states is equivalent to
 * comparing the 16 bytes big-endian — the form the canonical spelling prints. Lowercasing is
 * load-bearing rather than cosmetic: in ASCII `'0'–'9' < 'A'–'F' < 'a'–'f'`, so an id spelled in
 * upper case would sort below every lower-case one and this comparison would depend on how somebody
 * typed a GUID.
 */
function beats(challenger: PriceCandidate, holder: PriceCandidate): boolean {
  if (challenger.scope !== holder.scope) {
    return SPECIFICITY[challenger.scope] > SPECIFICITY[holder.scope];
  }

  if (challenger.effectiveFrom !== holder.effectiveFrom) {
    return challenger.effectiveFrom > holder.effectiveFrom;
  }

  return challenger.priceListId.toLowerCase() > holder.priceListId.toLowerCase();
}
